src/tests/conformance/post_handshake/validators.rs

```diff
@@ -55,3 +86,193 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+        .log_to_stdout(true)
+        .start(target.path(), NodeType::Stateless)
+        .await
+        .expect("unable to start stateful node");
+
+    let mut test_config = TestConfig::default();
+    test_config.synth_node_config.generate_new_keys = false;
+    let synth_node = SyntheticNode::new(&test_config).await;
+
+    synth_node
+        .connect(node.addr())
+        .await
+        .expect("unable to connect");
+
+    // 1. Setup keys & prefix.  Both master and signing key pairs have been previously generated.
+    let master_secret_hex = "8484781AE8EEB87D8A5AA38483B5CBBCCE6AD66B4185BB193DDDFAD5C1F4FC06";
+    // The master public key should be in the validators.txt file, in ~/.ziggurat/ripple/setup
+    let master_public_hex = "02ED521B8124454DD5B7769C813BD40E8D36E134DD51ACED873B49E165327F6DF2";
+    let master_secret_bytes = hex::decode(master_secret_hex).expect("unable to decode hex");
+    let master_public_bytes = hex::decode(master_public_hex).expect("unable to decode hex");
+    let master_secret_key =
+        SecretKey::from_slice(master_secret_bytes.as_slice()).expect("unable to create secret key");
+
+    let signing_secret_hex = "00F963180681C0D1D51D1128096B8FF8668AFDC41CBDED511D12D390105EFDDC";
+    let signing_public_hex = "03859B76317C8AA64F2D253D3547831E413F2663AE2568F7A17E85B283CC8861E4";
+    let signing_secret_bytes = hex::decode(signing_secret_hex).expect("unable to decode hex");
+    let signing_public_bytes = hex::decode(signing_public_hex).expect("unable to decode hex");
+    let signing_secret_key = SecretKey::from_slice(signing_secret_bytes.as_slice())
+        .expect("unable to create secret key");
+
+    // 2. Create manifest with sequence, public key, signing public key (without signatures)
+    assert_eq!(
+        master_public_bytes.len(),
+        PUBLIC_KEY_SIZE,
+        "invalid master public key length: {}",
+        master_public_bytes.len()
+    );
+    assert_eq!(
+        signing_public_bytes.len(),
+        PUBLIC_KEY_SIZE,
+        "invalid signing public key length: {}",
+        master_public_bytes.len()
+    );
+    let manifest = create_manifest(1, &master_public_bytes, &signing_public_bytes);
+
+    // 3. Sign the manifest with master secret key, get master signature
+    let master_signature_bytes =
+        sign_buffer_with_prefix(MANIFEST_PREFIX, &master_secret_key, &manifest);
+
+    // 4. Sign it with signing private key, get signature
+    let signature_bytes = sign_buffer_with_prefix(MANIFEST_PREFIX, &signing_secret_key, &manifest);
+
+    // 5. Create signed manifest with sequence, public key, signing public key, master signature, signature
+    let signed_manifest = sign_manifest(&manifest, &master_signature_bytes, &signature_bytes);
+
+    // 6. Create Validator blob.
+    let blob = create_validator_blob_json(&signed_manifest, master_public_hex);
+    let blob_b64 = base64::encode(&blob);
+    let blob_bytes = base64::decode(&blob_b64).expect("unable to decode a blob");
```

Comment:
Let's try to avoid doing a decode operation just after we encode something. We could instead copy the pre-encoded data into a separate variable before we do the encoding.
Or see my other suggestion below.

Reply:
done